ocimotel/repoclient.go

```diff
@@ -163,65 +164,85 @@ func (o *OciRepo) StartLayer() (string, error) {
 	}
 	defer resp.Body.Close()
 
-	if resp.StatusCode != 201 {
+	if resp.StatusCode != 202 {
 		return "", fmt.Errorf("Server returned an error %d", resp.StatusCode)
 	}
 
-	body, err := ioutil.ReadAll(resp.Body)
+	_, err = ioutil.ReadAll(resp.Body)
 	if err != nil {
 		return "", errors.Wrapf(err, "Error reading response body for %s", name)
 	}
 
-	var ret layerPostResult
-	err = json.Unmarshal(body, &ret)
-	if err != nil {
+	loc := resp.Header.Get("Location")
+	if loc == "" {
 		return "", errors.Wrap(err, "Failed decoding response")
 	}
 
-	return ret.Location, nil
+	return loc, nil
 }
 
 // @path is the uuid upload path returned by the server to our Post request.
 // @stream is the data source for the layer.
 // Return the digest and size of the layer that was uploaded.
-func (o *OciRepo) CompleteLayer(path string, stream io.Reader) (digest.Digest, int64, error) {
+func (o *OciRepo) CompleteLayer(path string, stream io.Reader, size int64) (digest.Digest, int64, error) {
 	uri := fmt.Sprintf("%s%s", o.url, path)
+	// using "chunked" upload
 	client := &http.Client{}
 	digester := sha256.New()
 	hashReader := io.TeeReader(stream, digester)
 	req, err := http.NewRequest("PATCH", uri, hashReader)
 	if err != nil {
 		return "", -1, errors.Wrap(err, "Failed opening Patch request")
 	}
+	req.ContentLength = size
+	req.Header.Set("Content-Type", "application/octet-stream")
+	req.Header.Set("Content-Range", fmt.Sprintf("%d-%d", 0, size))
 	resp, err := client.Do(req)
 	if err != nil {
 		return "", -1, errors.Wrapf(err, "Failed posting request %v", req)
 	}
 	defer resp.Body.Close()
-	if resp.StatusCode != 204 {
+	if resp.StatusCode != 202 {
 		return "", -1, fmt.Errorf("Server returned an error %d", resp.StatusCode)
 	}
 
 	ourDigest := fmt.Sprintf("%x", digester.Sum(nil))
-	uri = fmt.Sprintf("%s%s?digest=%s", o.url, path, ourDigest)
+	d := digest.NewDigestFromEncoded(digest.SHA256, ourDigest)
+	uri = fmt.Sprintf("%s%s?digest=%s", o.url, path, d.String())
 	req, err = http.NewRequest("PUT", uri, nil)
 	if err != nil {
 		return "", -1, errors.Wrap(err, "Failed opening Put request")
 	}
+	req.Header.Set("Content-Range", fmt.Sprintf("%d-%d", 0, size))
 	putResp, err := client.Do(req)
 	if err != nil {
 		return "", -1, errors.Wrapf(err, "Failed putting request %v", req)
 	}
 	defer putResp.Body.Close()
-	if putResp.StatusCode != 204 {
```

Comment:
Do we want to still accept 204, and just spit out an error?  Just wondering about any binaries which have been built with this library already.

Reply:
this concern should be applicable to only the ocimotel path? which we plan to replace end-to-end? is it still a valid concern?